### PR TITLE
Implement multiple attachments

### LINKED ISSUE #85: Support multiple attachments
You can now send up to 10 attachments with a message.

### PR DESCRIPTION
**Pull request by Flat**
Fixes #85

### CODE REVIEW

File: src/http/mod.rs [hunk 1]
```diff
@@ -70,6 +72,33 @@ pub enum LightMethod {
...
 
+/// Enum that allows a user to pass a `Path` or a `File` type to `send_files`
+pub enum FileOrPath {
+    /// Indicates that the `FileOrPath` is a `File`
+    File((File, String)),
+    /// Indicates that the `FileOrPath` is a `Path`
+    Path(PathBuf),
+}
+
+impl From<String> for FileOrPath{
```
Comment (zeyla) [inline src/http/mod.rs]: cs: space before `{`

File: src/http/mod.rs [hunk 2]
```diff
@@ -70,6 +72,33 @@ pub enum LightMethod {
     Put,
 }
 
+/// Enum that allows a user to pass a `Path` or a `File` type to `send_files`
+pub enum FileOrPath {
```
Comment (zeyla) [inline src/http/mod.rs]: Can we move this at the very bottom, where `GuildPagination` is?

File: src/http/mod.rs [hunk 3]
```diff
@@ -1403,6 +1433,55 @@ pub fn send_file<R: Read>(channel_id: u64, mut file: R, filename: &str, map: Jso
...
+
+    let mut request = Request::new(Method::Post, url)?;
+    request.headers_mut()
+        .set(header::Authorization(TOKEN.lock().unwrap().clone()));
+    request.headers_mut()
+        .set(header::UserAgent(constants::USER_AGENT.to_owned()));
+
+    let mut request = Multipart::from_request(request)?;
+    let mut file_num = String::from("0");
+    for file in files {
```
Comment (zeyla) [inline src/http/mod.rs]: cs: empty line before and after blocks

File: src/http/mod.rs [hunk 4]
```diff
@@ -1403,6 +1433,55 @@ pub fn send_file<R: Read>(channel_id: u64, mut file: R, filename: &str, map: Jso
...
+        match file.into() {
+            FileOrPath::File(f) => {
+                let (mut f, filename) = f;
+                request.write_stream(&file_num, &mut f, Some(&filename), None)?;
+            },
+            FileOrPath::Path(p) => {
+                request.write_file(&file_num, &p)?;
+            }, 
+        }
+        unsafe {
```
Comment (zeyla) [inline src/http/mod.rs]: cs: and here

File: src/http/mod.rs [hunk 5]
```diff
@@ -1403,6 +1433,55 @@ pub fn send_file<R: Read>(channel_id: u64, mut file: R, filename: &str, map: Jso
...
+            let vec = file_num.as_mut_vec();
+            vec[0] += 1;
+        }
+
```
Comment (zeyla) [inline src/http/mod.rs]: cs: remove this empty line

File: src/model/channel/channel_id.rs [hunk 6]
```diff
@@ -404,6 +408,78 @@ impl ChannelId {
         http::send_file(self.0, file, filename, map)
     }
 
+        /// Sends a file along with optional message contents. The filename _must_
```
Comment (zeyla) [inline src/model/channel/channel_id.rs]: cs: documentation is indented 4 spaces too much

File: src/model/channel/channel_id.rs [hunk 7]
```diff
@@ -404,6 +408,78 @@ impl ChannelId {
...
+    /// **Note**: Message contents must be under 2000 unicode code points.
+    ///
+    /// # Examples
+    ///
+    /// Send files with the paths `/path/to/file.jpg` and `/path/to/file2.jpg`
+    ///
+    /// ```rust,no_run
+    /// use serenity::model::ChannelId;
+    ///
+    ///
```
Comment (zeyla) [inline src/model/channel/channel_id.rs]: cs: only one empty line here

File: src/model/channel/channel_id.rs [hunk 8]
```diff
@@ -404,6 +408,78 @@ impl ChannelId {
...
+    /// let channel_id = ChannelId(7);
+    /// 
+    /// let paths = vec!("/path/to/file.jpg", "path/to/file2.jpg");
+    ///
+    /// let _ = channel_id.send_files(paths, |m| m.content("a file"));
+    /// ```
+    /// 
+    /// Send files using `File`
```
Comment (zeyla) [inline src/model/channel/channel_id.rs]:
```
Send files using `File`:
```

File: src/model/channel/group.rs [hunk 9]
```diff
@@ -293,11 +295,36 @@ impl Group {
...
+    ///
+    /// If the content of the message is over the above limit, then a
+    /// [`ClientError::MessageTooLong`] will be returned, containing the number
+    /// of unicode code points over the limit.
+    ///
+    /// [`ChannelId::send_file`]: struct.ChannelId.html#method.send_file
+    /// [`ClientError::MessageTooLong`]: ../client/enum.ClientError.html#variant.MessageTooLong
+    /// [Attach Files]: permissions/constant.ATTACH_FILES.html
+    /// [Send Messages]: permissions/constant.SEND_MESSAGES.html
+    pub fn send_files<F, T: Into<FileOrPath>>(&self, files: Vec<T>, f: F) -> Result<Message>
```
Comment (zeyla) [inline src/model/channel/group.rs]: This should be able to be inlined

File: src/model/channel/guild_channel.rs [hunk 10]
```diff
@@ -519,11 +521,36 @@ impl GuildChannel {
...
+    ///
+    /// If the content of the message is over the above limit, then a
+    /// [`ClientError::MessageTooLong`] will be returned, containing the number
+    /// of unicode code points over the limit.
+    ///
+    /// [`ChannelId::send_file`]: struct.ChannelId.html#method.send_file
+    /// [`ClientError::MessageTooLong`]: ../client/enum.ClientError.html#variant.MessageTooLong
+    /// [Attach Files]: permissions/constant.ATTACH_FILES.html
+    /// [Send Messages]: permissions/constant.SEND_MESSAGES.html
+    pub fn send_files<F, T: Into<FileOrPath>>(&self, files: Vec<T>, f: F) -> Result<Message>
```
Comment (zeyla) [inline src/model/channel/guild_channel.rs]: ^ and here

File: src/model/channel/mod.rs [hunk 11]
```diff
@@ -288,11 +290,36 @@ impl Channel {
...
+    ///
+    /// If the content of the message is over the above limit, then a
+    /// [`ClientError::MessageTooLong`] will be returned, containing the number
+    /// of unicode code points over the limit.
+    ///
+    /// [`ChannelId::send_file`]: struct.ChannelId.html#method.send_file
+    /// [`ClientError::MessageTooLong`]: ../client/enum.ClientError.html#variant.MessageTooLong
+    /// [Attach Files]: permissions/constant.ATTACH_FILES.html
+    /// [Send Messages]: permissions/constant.SEND_MESSAGES.html
+    pub fn send_files<F, T: Into<FileOrPath>>(&self, files: Vec<T>, f: F) -> Result<Message>
```
Comment (zeyla) [inline src/model/channel/mod.rs]: ^ and here

File: src/model/channel/private_channel.rs [hunk 12]
```diff
@@ -239,11 +241,36 @@ impl PrivateChannel {
...
+    ///
+    /// If the content of the message is over the above limit, then a
+    /// [`ClientError::MessageTooLong`] will be returned, containing the number
+    /// of unicode code points over the limit.
+    ///
+    /// [`ChannelId::send_file`]: struct.ChannelId.html#method.send_file
+    /// [`ClientError::MessageTooLong`]: ../client/enum.ClientError.html#variant.MessageTooLong
+    /// [Attach Files]: permissions/constant.ATTACH_FILES.html
+    /// [Send Messages]: permissions/constant.SEND_MESSAGES.html
+    pub fn send_files<F, T: Into<FileOrPath>>(&self, files: Vec<T>, f: F) -> Result<Message>
```
Comment (zeyla) [inline src/model/channel/private_channel.rs]: ^ and here

File: src/model/channel/channel_id.rs [hunk 13]
```diff
@@ -404,6 +408,78 @@ impl ChannelId {
...
+    /// An embed can _not_ be sent when sending a file. If you set one, it will
+    /// be automatically removed.
+    ///
+    /// The [Attach Files] and [Send Messages] permissions are required.
+    ///
+    /// **Note**: Message contents must be under 2000 unicode code points.
+    ///
+    /// # Examples
+    ///
+    /// Send files with the paths `/path/to/file.jpg` and `/path/to/file2.jpg`
```
Comment (zeyla) [inline src/model/channel/channel_id.rs]: Colon at the end here

File: src/model/channel/channel_id.rs [hunk 14]
```diff
@@ -404,6 +408,78 @@ impl ChannelId {
...
+    /// ```rust,no_run
+    /// use serenity::model::ChannelId;
+    /// use std::fs::File;
+    ///
+    /// let channel_id = ChannelId(7);
+    ///
+    /// let f1 = File::open("my_file.jpg").unwrap();
+    /// let f2 = File::open("my_file2.jpg").unwrap();
+    ///
+    /// let files = vec!((f1, "my_file.jpg"), (f2, "my_file2.jpg"));
```
Comment (zeyla) [inline src/model/channel/channel_id.rs]: `vec![]` here too

File: src/http/mod.rs [hunk 15]
```diff
@@ -70,6 +72,33 @@ pub enum LightMethod {
...
+impl<'a> From<&'a str> for FileOrPath {
+    fn from(s: &'a str) -> FileOrPath {
+        FileOrPath::Path(PathBuf::from(s))
+    }
+}
+
+impl<'a> From<(File, &'a str)> for FileOrPath {
+    fn from(f: (File, &str)) -> FileOrPath {
+        let (file, filename) = f;
+        FileOrPath::File((file, filename.to_owned()))
```
Comment (zeyla) [inline src/http/mod.rs]: Should just access the tuplestruct's values directly instead of pattern matching into bindings.

File: src/http/mod.rs [hunk 16]
```diff
@@ -1403,6 +1433,55 @@ pub fn send_file<R: Read>(channel_id: u64, mut file: R, filename: &str, map: Jso
...
+    };
+
+    let mut request = Request::new(Method::Post, url)?;
+    request.headers_mut()
+        .set(header::Authorization(TOKEN.lock().unwrap().clone()));
+    request.headers_mut()
+        .set(header::UserAgent(constants::USER_AGENT.to_owned()));
+
+    let mut request = Multipart::from_request(request)?;
+    let mut file_num = String::from("0");
```
Comment (zeyla) [inline src/http/mod.rs]: `"0".to_owned()`

File: src/http/mod.rs [hunk 17]
```diff
@@ -1403,6 +1433,55 @@ pub fn send_file<R: Read>(channel_id: u64, mut file: R, filename: &str, map: Jso
...
+    for file in files {
+        match file.into() {
+            FileOrPath::File(f) => {
+                let (mut f, filename) = f;
```
Comment (zeyla) [inline src/http/mod.rs]: We should instead move this up and do `FileOrPath::File((mut f, filename)) => {`

File: src/model/channel/channel_id.rs [hunk 18]
```diff
@@ -404,6 +408,78 @@ impl ChannelId {
...
+    ///
+    /// Send files with the paths `/path/to/file.jpg` and `/path/to/file2.jpg`
+    ///
+    /// ```rust,no_run
+    /// use serenity::model::ChannelId;
+    ///
+    ///
+    /// let channel_id = ChannelId(7);
+    /// 
+    /// let paths = vec!("/path/to/file.jpg", "path/to/file2.jpg");
```
Comment (zeyla) [inline src/model/channel/channel_id.rs]: `vec![]` syntax is preferred

**Comment by zeyla** [comment]
Oh, and needs a rebase too.

File: src/http/mod.rs [hunk 19]
```diff
@@ -1403,6 +1433,55 @@ pub fn send_file<R: Read>(channel_id: u64, mut file: R, filename: &str, map: Jso
...
+        }
+        unsafe {
+            let vec = file_num.as_mut_vec();
+            vec[0] += 1;
+        }
+        
+    }
+
+    for (k, v) in map {
+        let _ = match v {
```
Comment (acdenisSK) [inline src/http/mod.rs]: This let binding is useless

**Comment by zeyla** [comment]
Looks good, thanks!

**Pull request merged**